### PR TITLE
Snowflake: Handle FLATTEN() table function aliases correctly in L025, L027, L028

### LINKED ISSUE #3178: Snowflake: Interaction between L025 and L028 corrupts SQL
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqfluff fix 
```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
```
Produces:
```
L:   3 | P:   5 | L028 | Unqualified reference 'value' found in single table select.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.

SELECT
  r.rec:foo::string AS foo
  , r.value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result)
;
```
The issue here is it doesn't recognise the `LATERAL FLATTEN` as a table entity. Replacing the `LATERAL FLATTEN` with `bar` gives

```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, bar x
;

L:   4 | P:  21 | L025 | Alias 'x' is never used in SELECT statement.
```

### Expected Behaviour

LATERAL FLATTEN should be treated like a table

### Observed Behaviour

-

### How to reproduce

-

### Dialect

Snowflake

### Version

0.13.0

### Configuration

Unaffected by config

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #3178

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:91]: Because this was missing, [this code](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/rules/analysis/select.py#L137L138) was not executing, preventing the `FLATTEN()` call from being recognized as returning a table.

**Comment by WittierDinosaur** [comment]
So there are actually a whole bunch of table functions in snowflake: 
https://docs.snowflake.com/en/sql-reference/functions-table.html
I'll leave it up to you whether you think adding them all is in-scope of this ticket. Happy to approve if you want to do this bugfix independently

**Comment by barrywhart** [comment]
@WittierDinosaur: Let's create a separate issue for the other functions. Also, even with this fix, the query is still being modified by other rules, although it's now leaving the `x` alias in place. I'll look at avoiding these other "fixes" as well. I'm guessing other rules are still ignoring the `FLATTEN` as a "table", even though the core function works now.
```
SELECT
    foo.bar.rec:foo::string AS foo,
    foo.bar.value:bar::string AS bar
FROM foo.bar, LATERAL FLATTEN(input => rec:result) AS x;
```

**Comment by WittierDinosaur** [comment]
Cool, scope-wise then. I think stopping the query being modified is in scope, I'll make another issue for further table function support

**Review by tunetheweb — APPROVED** [review]

**Comment by barrywhart** [comment]
I'm putting this PR back in "draft" status. The Snowflake `FLATTEN()` function is a bit different than the similar `UNNEST()` function in BigQuery, which we added support for in PR #722.

`UNNEST()` returns an ambiguous thing that is both a table and a column. On the other hand, I think `FLATTEN()` returns a true table.

File: src/sqlfluff/rules/L025.py
Comment (barrywhart) [inline src/sqlfluff/rules/L025.py:93]: Important decision here -- we're not going to remove unused aliases for table functions, even if they're unused. I'm not sure this is the perfect solution, but it seems good enough for now.

**Comment by tunetheweb** [comment]
> `UNNEST()` returns an ambiguous thing that is both a table and a column. On the other hand, I think `FLATTEN()` returns a true table.

If it is a true table, and the alias isn’t used, then isn’t it right to remove it?

As per discussion in #3178 @WittierDinosaur was happy L025 was flagging - just not happy L028 was flagging in `fix` even though not in `lint`.

File: src/sqlfluff/core/rules/analysis/select.py
Comment (barrywhart) [inline src/sqlfluff/core/rules/analysis/select.py:120]: This is unrelated to the overall fix. It looks like a typo -- the intent here was to avoid adding duplicate entries to `table_aliases`.

**Comment by barrywhart** [comment]
@WittierDinosaur: This issue required some tricky, fairly deep changes. Would you like to review and/or test this on your actual SQL?

**Comment by barrywhart** [comment]
@tunetheweb: Can you review this again? It has grown and changed a lot since your initial review.

**Comment by barrywhart** [comment]
@tunetheweb

Re:
>If it is a true table, and the alias isn’t used, then isn’t it right to remove it?

Please see the comments I added to L025. It acts like a table in the sense that it has multiple, named columns. It is **not** like a table in that if you remove the alias, there's no underlying table name you can use instead of the alias to access that data.

Once the alias is removed, L028 gets confused. It thinks there's only one possible source for columns, so it makes a bad fix. Do you have an alternate proposal that will make L025, L027, and L028 do something reasonable and not interact or interfere with one another in strange ways? Look back at the discussion between you and @WittierDinosaur on the related issue -- very confusing interaction. I think this is a workable solution, and I'm reluctant to change "just one thing" which re-introduces some of those other problems. What's "right" in isolation (one rule) isn't necessarily going to work well in combination with the other rules. 

If we don't have other ideas how to make all these things work well, I propose let's go with the approach here (which I believe is safe, even if it allows unused aliases to linger). Better to leave the SQL a little unclean than to break it! We can always revisit this later.

**Comment by barrywhart** [comment]
I do have one other idea -- I'll branch off this branch, then create a second PR for consideration (if it works!).

**Comment by barrywhart** [comment]
@tunetheweb: The other idea worked out so well that I just went ahead and updated this PR instead. 🙌 😎

File: test/fixtures/rules/std_rule_cases/L027.yml
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L027.yml:282]: Should we have a third test here?

```suggestion
test_fail_table_plus_flatten_snowflake_1:
  # FLATTEN() returns a table, thus there are two tables, thus lint failure
  fail_str: |
    SELECT
        r.rec:foo::string AS foo,
        value:bar::string AS bar
    FROM foo.bar AS r, LATERAL FLATTEN(input => rec:result) AS x
  configs:
    core:
      dialect: snowflake

test_fail_table_plus_flatten_snowflake_2:
  # FLATTEN() returns a table, thus there are two tables, thus lint failure,
  # even though there's no alias provided for FLATTEN()
  fail_str: |
    SELECT
        r.rec:foo::string AS foo,
        value:bar::string AS bar
    FROM foo.bar AS r, LATERAL FLATTEN(input => rec:result)
  configs:
    core:
      dialect: snowflake

test_fail_table_plus_flatten_snowflake_3:
  # FLATTEN() returns a table, thus there are two tables, thus lint failure,
  # even though there's no alias provided for FLATTEN()
  pass_str: |
    SELECT
        r.rec:foo::string AS foo,
        x.value:bar::string AS bar
    FROM foo.bar AS r, LATERAL FLATTEN(input => rec:result) AS x
  configs:
    core:
      dialect: snowflake
```

**Review by tunetheweb — APPROVED** [review]
LGTM - one small suggestion for an additional test case

File: test/fixtures/rules/std_rule_cases/L027.yml
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L027.yml:282]: Sure! I'll add that. I assume you want it to be a passing test case, in spite of "fail" in the suggested name and comment?

Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L027.yml]: Added!

Comment (tunetheweb) [reply test/fixtures/rules/std_rule_cases/L027.yml]: Lazy copy and paste :-)

**Pull request merged**
